batch_requests.go

```diff
@@ -0,0 +1,150 @@
+// You can add upto 20 BatchItems to a BatchRequest
+func (br *batchRequest) AppendBatchItem(reqInfo abstractions.RequestInformation) (*batchItem, error) {
+	if len(br.Requests) > 19 {
+		return nil, errors.New("Batch items limit exceeded. BatchRequest has a limit of 20 batch items")
+	}
+
+	batchItem, err := toBatchItem(reqInfo)
+	if weGotAnError(err) {
+		return nil, err
+	}
+
+	br.Requests = append(br.Requests, batchItem)
+	return batchItem, nil
+}
+
+// DependsOnItem creates a dependency chain between BatchItems.If A depends on B, then B will be sent before B
+func (bi *batchItem) DependsOnItem(item batchItem) {
+	// DependsOn is a single value slice
+	bi.DependsOn = []string{item.Id}
+}
+
+// NewBatchRequest creates an instance of BatchRequest
+func NewBatchRequest() *batchRequest {
+	return &batchRequest{}
+}
+
+func toBatchItem(requestInfo abstractions.RequestInformation) (*batchItem, error) {
+	url, err := requestInfo.GetUri()
+	if weGotAnError(err) {
+		return nil, err
+	}
+
+	var body map[string]any
+	json.Unmarshal(requestInfo.Content, &body)
+
+	return &batchItem{
+		Id:        uuid.NewString(),
+		Method:    requestInfo.Method.String(),
+		Body:      body,
+		Headers:   requestInfo.Headers,
+		Url:       url.Path,
+		DependsOn: make([]string, 0),
+	}, nil
+}
+
+func (r batchRequest) toJson() ([]byte, error) {
+	return json.Marshal(r)
+}
+
+func getBaseUrl(adapter abstractions.RequestAdapter) (*url.URL, error) {
+	return url.Parse(adapter.GetBaseUrl())
+}
+
+func buildRequestInfo(jsonBody []byte, baseUrl *url.URL) *abstractions.RequestInformation {
+	requestInfo := abstractions.NewRequestInformation()
+
+	requestInfo.SetStreamContent(jsonBody)
+	requestInfo.Method = abstractions.POST
+	requestInfo.SetUri(*baseUrl)
+	requestInfo.Headers = map[string]string{
+		"Content-Type": "application/json",
+	}
+
+	return requestInfo
+}
+
+func sendBatchRequest(requestInfo *abstractions.RequestInformation, adapter abstractions.RequestAdapter) (BatchResponse, error) {
+	var res BatchResponse
+
+	// we don't care about SendAsync's return value because we don't use it and it is a workaround
+	// on SendAsync's insistence to have a parsable node as a return value
+	_, err := adapter.SendAsync(requestInfo, nil, func(response any, errorMappings abstractions.ErrorMappings) (any, error) {
+		resp, ok := response.(*http.Response)
+		if !ok {
+			return nil, errors.New("Response type assertion failed")
+		}
+
+		if status := resp.StatusCode; status >= 400 && status <= 600 {
+			return res, fmt.Errorf("Request failed with status: %d", status)
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if weGotAnError(err) {
+			return res, err
+		}
+
+		json.Unmarshal(body, &res)
+
+		// returning a Noop Parsable here because SendAsync type asserts HandlerFunc's return value
+		return internal.NewNoOpParsable(), nil
+	}, nil)
+
+	if weGotAnError(err) {
+		return res, err
+	}
+
+	return res, nil
+}
+
+type batchRequest struct {
+	Requests []*batchItem `json:"requests"`
```

Comment:
any specific reason why this one is an array of pointers and the other one just an array?

Reply:
Yes, because of the DependsOn relationship which is specified after a request has been added.

Reply:
Would there be any downside to aligning both to pointers?

Reply:
Consistency

Reply:
Returning a pointer or passing pointers around has significance means that the struct pointed to can be mutated and of course there's a cost to that. I wouldn't recommend doing so just for consistency.

Reply:
While I agree that a pointer signals mutation, I do think you should review the assumption about cost, especially on "large structs".
https://stackoverflow.com/questions/23542989/pointers-vs-values-in-parameters-and-return-values
https://stackoverflow.com/questions/26906323/golang-pointer-performance

Nonetheless, the array here is already a pointer to begin with, let's try to take the problem the other way around, is there specific reasons why we need pointers on the Requests? (on top of the slice)

Reply:
Yes the reason why Requests is a slice of pointers to batchItems is because the batchItems can change after they have been added to the array. A batch item can have it dependsOn property updated after it has been added to the Requests slice.